src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -67,6 +67,358 @@ public static ControlFlowGraph Create(IBlockOperation body)
+                //                      For example, strings, decimal, datetime constants are not primitive value
+                //                      type constants.
+                if (operation.ConstantValue.HasValue)
+                {
+                    return false;
+                }
+
+                // Treat an exception assignment to a local or a parameter as non-throwing
+                if (operation.Kind == OperationKind.SimpleAssignment)
+                {
+                    var assignment = (ISimpleAssignmentOperation)operation;
+
+                    if (assignment.Value.Kind == OperationKind.CaughtException &&
+                        (assignment.Target.Kind == OperationKind.LocalReference || assignment.Target.Kind == OperationKind.ParameterReference))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+
+            void followBranch(BasicBlock current, BasicBlock.Branch branch, ref bool canThrow)
+            {
+                switch (branch.Kind)
+                {
+                    case BasicBlock.BranchKind.None:
+                    case BasicBlock.BranchKind.ProgramTermination:
+                    case BasicBlock.BranchKind.StructuredExceptionHandling:
+                        Debug.Assert(branch.Destination == null);
+                        return;
+
+                    case BasicBlock.BranchKind.Throw:
+                    case BasicBlock.BranchKind.ReThrow:
+                        Debug.Assert(branch.Destination == null);
+                        canThrow = true;
+                        return;
+
+                    case BasicBlock.BranchKind.Regular:
+                    case BasicBlock.BranchKind.Return:
+                        Debug.Assert(branch.Destination != null);
+
+                        if (stepThroughFinally(current.Region, branch.Destination))
+                        {
+                            toVisit.Add(branch.Destination);
+                        }
+
+                        return;
+
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(branch.Kind);
+                }
+            }
+
+            // Returns whether we should proceed to the destination after finallies were taken care of.
+            bool stepThroughFinally(ControlFlowGraph.Region region, BasicBlock destination)
+            {
+                int destinationOrdinal = destination.Ordinal;
+                while (region.FirstBlockOrdinal > destinationOrdinal || region.LastBlockOrdinal < destinationOrdinal)
+                {
+                    ControlFlowGraph.Region enclosing = region.Enclosing;
+                    if (region.Kind == ControlFlowGraph.RegionKind.Try)
+                    {
+                        if (enclosing.Kind == ControlFlowGraph.RegionKind.TryAndFinally)
```

Comment:
Nit: Could just combine this if and the enclosing if, feels like less nesting would be more readable.